drivers/include: add header definition for wdt

fjmolinas opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->

This PR adds header files defining  a proposition for a common wdt api. It's based on implementations discussed in #11520, #7374 and #11252,. This PR aims at unifying the 3 implementations and achieving an implementation that covers  stm32, sam0 and cc2538.

4 functions are defined:

```
void wdt_start(void);
void wdt_stop(void);
void wdt_kick(void);
int wdt_setup(uint32_t min_time, uint32_t max_time, wdt_cb_t wdt_cb, void* arg);
```

In many cases (e.g. stm32) the wdt can't be stopped so this function would have no effect. In cases where a callback is not possible these can be defined as  NULL. 'min_time' and 'max_time' must must be defined with valid values for windowed timer, while 'min_time=0' would equal normal wdg operation.

Range limits for the WDT should be known at compile time although they could be different for normal operation and windowed operation. 'WDT_TIME_LOWER_BOUND' et al should be defined per CPU.

```
/**
 * @name    WDT upper and lower bound times in us
 * @{
 */
#define WDT_TIME_LOWER_BOUND            Y
#define WDT_TIME_UPPER_BOUND            X
#define WWDT_TIME_LOWER_BOUND           YY
#define WWDT_TIME_UPPER_BOUND           XX
/** @} */
```

Specially interested in thoughts @geith and @benpicco have on the subject since they implemented similar similar header definitions. Since this header definition took from all of their definitions it should be able to work for all referenced implementations.

### Testing procedure

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->

Testing should be addressed in the related PR's.

### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

Implementations following this api in #11520 and #11252.

Based on discussion in #11520, #7374 and #11252

kaspar030 commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,77 @@
+ * @param[in] min_time       lower bound for windowed watchdog in us
+ * @param[in] max_time       upper bound for windowed watchdog in us, time before
+ *                           reset for normal watchdog
+ * @param[in] wdt_cb         wdt callback
```
We could default to reboot, of no callback is given.

It might make sense to split this into to:

```
void wdt_setup_reboot(uint32_t min_time, uint32_t max_time);
void wdt_setup_callback(uint32_t min_time, uint32_t max_time, wdt_cb_t wdt_cb, void* arg);
```

1. the latter can be guarded by a feature (e.g., periph_wdt_callback), should it not be available or desired
2. even if available but not used, the actual code will always have to reserve the 8 bytes of memory.

benpicco replied:
At least on sam0 the Watchdog will always trigger a reboot, even when the early warning interrupt (callback) is enabled.

The callback is optional, when set it will be called a set time (e.g. one period) before the watchdog timer expires.

fjmolinas replied:
In the case of stm32 the callback can feed/kick/reset de watchdog so a reboot is not the necessary outcome, it would make sense to me to split them having a separate function to setup the callback, which could be guarded.

kaspar030 replied:
Well, we might have to make some concessions to a generic API here. The behaviour needs to be *the same* on all platforms implementing the interface. That's why I think it makes sense to split reboot and callback functionality.

How do other platforms handle callbacks/interrupts from the WDT?

(skimming datasheets)

STM32F4 seems to have a reset-only watchdog (IWDG) and a windowing one with reset + early warning interrupt.

nrf52 also has an early warning interrupt, but the datasheet says "Once the TIMEOUT event has been generated, the impending watchdog reset will always be effectuated.", suggesting that the watchdog cannot be reset from the interrupt.

kaspar030 replied:
we could define the early warning interrupt in a way that all platforms support (e.g., it'll be triggered but cannot reset the watchdog), and emulate it using xtimer for platforms not supporting it...

Anyhow, I suggest we go with only ```wdt_setup_reset()``` for now. It'll be ok for 90% of the users, and we can extend the API later. What do you think?

benpicco commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,77 @@
+/**
+ * @brief    Reset the watchdog timer counter, delay system reset
+ */
+void wdt_kick(void);
```
I think that analogy is needlessly cruel. Instead of perpetually beating the watchdog unconscious I think we should rather `wdt_feed` it and have it reset the system when it starves.

We could also `wdt_pet` it, but I think the behavior when stopped being petted is not universal among all watchdogs.

fjmolinas replied:
I wasn't actually trying to make an analogy when naming it 'wdg_kick', I was thinking of kick as delay/postpone. But I get your point, I don't mind changing it to feed (although starvation doesn't seem to me like a nice outcome either hahaha)

kaspar030 replied:
Hmm, "kick" seems to be the terminus used on the wikipedia [page](https://en.wikipedia.org/wiki/Watchdog_timer). It is mentioned as an alternative to "restart".

fjmolinas replied:
In other references.

> Kicking the dog
> 
> 
> The process of restarting the watchdog timer's counter is sometimes called "kicking the dog." The appropriate visual metaphor is that of a man being attacked by a vicious dog. If he keeps kicking the dog, it can't ever bite him. But he must keep kicking the dog at regular intervals to avoid a bite. Similarly, the software must restart the watchdog timer at a regular rate, or risk being restarted.

It does seem to be the defacto word used in must watchdog implementations so I would keep it.

fjmolinas commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,77 @@
+ * @brief    Set up the wdt timer, only use max_time if normal operation
+ *           set min_time and max_time for windowed timer.
+ *
+ * @param[in] min_time       lower bound for windowed watchdog in us
```
Looking at datasheet for other platforms max timeout can vary. It goes up to some seconds for ATMEGA, STM32. Hours for MSP430, nd for KINETIS since it has a 32bit counter running from the LPO (1Khz) the max timeout can be set to many days. NRF can go up to 50hrs approx. For ESP I'm not very clear... seems to depend if its clocked from the RTC or BUS clock and goes between ms and s.

In that case using uin32_t for us would be a constraint since the max achievable time would be a little more than 1 hour and since platforms that can go under ms only go around 100us, it woyld make sense to use ms as the base unit instead of us.

benpicco replied:
I really doubt anyone would need µs precision for the watchdog.
Using ms here really seems like the most sensible approach.

fjmolinas commented:
@benpicco @kaspar030 I have pushed a couple of updates:

- changed units to ms
- I added 'wdt_setup_callback(uint32_t min_time, uint32_t max_time, wdt_cb_t wdt_cb, void* arg)' guarded feature. Since @benpicco implementation is using this I think it makes sense adding it right away. But if its not supported it will save a couple of bytes as @kaspar030 pointer out.
- extended documentation
- kept for now wdt_kick(), seem to be the most common term in different wdt implementations.

What do you think?

benpicco commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,169 @@
+ * @param[in] arg            optional argument which is passed to the
+ *                           callback, can be NULL
+ */
+void wdt_setup_callback(uint32_t min_time, uint32_t max_time, wdt_cb_t wdt_cb, void* arg);
```
I think this can be misleading: The watchdog will still trigger a reboot and behave just like with `wdt_setup_reboot`, the only difference being that `wdt_cb` is executed before the watchdog timer expires.

Having two functions for it make it seem like `wdt_setup_callback` would cause the callback to be called *instead* of the reboot, not in addition to it.

Imho one `wdt_setup(uint32_t min_time, uint32_t max_time, wdt_cb_t wdt_cb, void* arg)` should be enough. `wdt_cb` can be `NULL` if it's not needed.

Also it would be good if the function can return an error code if an invalid configuration is requested.

kaspar030 replied:
> I think this can be misleading: The watchdog will still trigger a reboot and behave just like with wdt_setup_reboot, the only difference being that wdt_cb is executed before the watchdog timer expires.

Maybe call it ```wdt_setup_reset_with_warning()```?

BTW, we need an additional define for the time difference between warning callback and actual reset.
 
> Having two functions for it make it seem like `wdt_setup_callback` would cause the callback to be called _instead_ of the reboot, not in addition to it.

If this is one function, the functionality cannot be guarded properly. When supported, all callback handling code must be compiled in, even if unused.
In practice, ```wdt_setup_reset_with_warning()``` will probably look like

```
wdt_setup_reset_with_warning() {
 wdt_setup_reset();
// set callback
}
```

> Also it would be good if the function can return an error code if an invalid configuration is requested.

Would be good for what? Can we get away with asserts and log messages?

```
if (wdt_setup_reset() == EINVAL) {
  // now what? set a flag disabling wdt_kick()?
  // let the user know?
}
```

benpicco replied:
I'm fine with `wdt_setup_reset_with_warning()` if that means that the compiler can LTO away those pesky 8 bytes if not used. 

Log messages on the other hand tend to be larger than 8 byte (yea, ROM hurts a little less than RAM, but still), but I guess we are good when we enable them when `DEVELHELP` is set.

fjmolinas replied:
> I think this can be misleading: The watchdog will still trigger a reboot and behave just like with wdt_setup_reboot, the only difference being that wdt_cb is executed before the watchdog timer expires.

> BTW, we need an additional define for the time difference between warning callback and actual 
reset.

For sam21 its configurable in normal mode but matches the start of the window in windowed mode (which seems useless to me), for nrf52 its a different configurable value and for stm32 its fixed at when the downcounter hits 0x40. A define would work for stm32, nrf52 but would not work in the case of windowed operation for sam21, but proper DEBUG messages could state this behavior. 

> Having two functions for it make it seem like wdt_setup_callback would cause the callback to be called instead of the reboot, not in addition to it.

Splitting into two functions makes sense to me since we really have very different behaviors for the callback according to each platform, while normal operation is pretty uniform.

> Can we get away with asserts and log messages?

+1

aabadie commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,169 @@
+ * platforms). 
+ *
+ * This basic interface defines two operation modes NORMAL and WINDOWED. In
+ * NORMAL operation if the WDT counter isn't reset before TIMEOUT than a
```
s/than/then/

aabadie commented:
I have 2 comments/questions:
- on nrf52, one can disable or keep the watchdog running during sleep/debug sessions. I guess other architectures also provide this kind of behaviors. How could this be handled by the current design ? Maybe using overridable defines at CPU levels and let the peripheral driver deal with it ? or at HAL level ?
- my second comment is about multiple watchdog management on the same CPU. still on nrf52, one can configure up to 8 watchdog "reload requests". STM32 provides 2 types of watchdog on the same CPU (independent and windowed). If I'm not mistaken, the current design doesn't allow to support such features.

aabadie commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,169 @@
+ * can't be stopped so verify the specific platform implementation and (some
+ * only pause/stop it in some operation modes).
+ *
+ * Since WDT upper and lower bound are a per platform constraint they can be
```
s/bound/bound**s**/

aabadie commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,169 @@
+ * only pause/stop it in some operation modes).
+ *
+ * Since WDT upper and lower bound are a per platform constraint they can be
+ * definend in periph_cpu.h and should be known at compile time. A distinction
```
s/definend/defined/

fjmolinas commented:
@aabadie @benpicco @kaspar030 

I pushed new changes taking your comments into account.

Regarding naming, Changed `wdt_setup_callback()` to `wdt_setup_reboot_with_callback()` so its clear it is doing both operations.

For the WDT callback I think the clearer way is adding a paramter `cb_time` to `wdt_setup_reboot_with_callback()`. This will make the api clearer for all use cases. 

* If fixed value: `assert(cb_time==0)`
* If configurable value it can be set as `cb_time`
* If configurable within certain values `assert(cb_time==OPTION_A || cb_time==OPTION_B ...)`
* If conflicts with min_time (samd): `assert(min_time != cb_time!=0)`

```
/**
 * @brief    Set up the wdt timer, only use max_time if NORMAL operation
 *           set min_time and max_time for WINDOWED timer.
 *
 * @param[in] min_time       lower bound for WINDOWED watchdog in ms, has to be 0
 *                           for NORMAL watchdog.
 * @param[in] max_time       upper bound for WINDOWED watchdog in ms, time before
 *                           reset for NORMAL watchdog.
 * @param[in] cb _time       time before wdt_cb is executed, must.
 * @param[in] wdt_cb         wdt callback, can be NULL
 * @param[in] arg            optional argument which is passed to the
 *                           callback, can be NULL
 */
void wdt_setup_reboot_with_callback(uint32_t min_time, uint32_t max_time,
                             uint32_t cb_time, wdt_cb_t wdt_cb, void* arg);
```

> on nrf52, one can disable or keep the watchdog running during sleep/debug sessions. I guess other architectures also provide this kind of behaviors. How could this be handled by the current design ? Maybe using overridable defines at CPU levels and let the peripheral driver deal with it ? or at HAL level ?

@aabadie This can indeed be the case. Overridable defines should do the trick but lets leave this for a future PR to try to limit the discussion on this PR, what do you think?

> my second comment is about multiple watchdog management on the same CPU. still on nrf52, one can configure up to 8 watchdog "reload requests". STM32 provides 2 types of watchdog on the same CPU (independent and windowed). If I'm not mistaken, the current design doesn't allow to support such features.

@aabadie  A choice has to be made between supporting all features and a simple API. From discussions in #7374 the choice was going for a simple one and not supporting all features. So I think a single WDT has to the way to go (unless in the feature someone puts forward a use case for multiple ones). What do you think?

kaspar030 commented:
> For the WDT callback I think the clearer way is adding a paramter `cb_time` to `wdt_setup_reboot_with_callback()`. This will make the api clearer for all use cases.

Well, that time is fixed on a couple of architectures I looked at. Maybe rather go with a define here?

We should make sure that applications time the warning callback so that the warning is issued at the watchdog timeout, and the reboot is ```WDT_WARNING_PERIOD``` after. Or the other way around. Just consistent. :)

When a callback is set, what happens if a minimum is set and then the wdt is kicked too early?

fjmolinas commented:
> Well, that time is fixed on a couple of architectures I looked at. 

Not for SAM21, its configurable although in WINDOWED mode it equals to the lower bound. But this makes the api confusing (to me at least), since in normal mode min_time is actually cb_time and in windowed they are equal, having them separate makes it easier to handle. 

For EFM32  EWI can occur at 25%, 50%, or 75% of the timeout.

> Maybe rather go with a define here?

Also if it was a define how do you handle it? In the case of STM32 its a fix counter value, but doesn't translate into `ms` it depends on the prescaler and reload value. In NRF52 its just a couple of clock cycles before reboot. So it can't actually be a define, nor in `ms` units at least. The defines would need to have different units per platform, which makes it a little messy to me. What do you think?
 
> We should make sure that applications time the warning callback so that the warning is issued at the watchdog timeout, and the reboot is ```WDT_WARNING_PERIOD``` after. Or the other way around. Just consistent. :)

@benpicco can correct me here, but this would make the callback unusable in windowed mode. Sould we the define  windowed and callback as features, that way they can conflict each other.

> When a callback is set, what happens if a minimum is set and then the wdt is kicked too early?

Not sure for all platforms. But for STM32 the callback isn't executed, there is an immediate reboot.

benpicco commented:
> We should make sure that applications time the warning callback so that the warning is issued at the watchdog timeout, and the reboot is `WDT_WARNING_PERIOD` after. Or the other way around. Just consistent. :)

The first option sounds like it's trying to be too clever for it's own good ;)
Having the callback  `WDT_WARNING_PERIOD` *before* the timeout sounds about right.
And if `WDT_WARNING_PERIOD` = `max_time` - `min_time` than that's fine.
 
> When a callback is set, what happens if a minimum is set and then the wdt is kicked too early?

As I understand it, the warning interrupt is useful if you run a completely event-based system.
There is no periodic loop to 'kick' the watchdog, but instead you have a periodic watchdog interrupt that does the kicking or schedules the dirty work within your event system.

If you kick the dog too early than that's an error and the system should be reset.

fjmolinas commented:
@kaspar030 ping!

fjmolinas commented:
@kaspar030 @benpicco I still think having `cb_time` is more flexible and would define the callback behavior better for samd. In the case of stm32 it can't be specified in `us`, the actual time will depend on the prescaler and reload value set for the WDT counter. 

If you are okay with having a DEBUG message stating this behavior and overruling `WDT_WARNING_PERIOD` and a conflicting future for `wdt_windowed` and `wdt_callback`. then a define would be ok (although not ideal). I'm not going to block this because of that (specially since I didn't actually implement the callback part yet). If you both agree with a define for  `WDT_WARNING_PERIOD` then I can push those changes so we can move ahead.

benpicco commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,177 @@
+ * |---------MIN_TIME-------|-----------WINDOW----------|
+ *
+ * |------------VALID_TIME-----------|
+ *                                   ^
```
should be
```
* |---------MIN_TIME-------|-----------WINDOW----------|
*
*                          |----------VALID_TIME------|
*                                                     ^
*                                                     |
*                                                 wdt_kick()
```

benpicco commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,177 @@
+ *
+ *    // MIN_TIME < VALID_TIME < MAX_TIME
+ *    while (1) {
+ *        xtimer_usleep(VALID_TIME);
```
```
 *        xtimer_usleep(MIN_TIME);
```

benpicco replied:
this is still wrong

fjmolinas replied:
Why? The window is `[MIN_TIME, MAX_TIME]`. If `kick()` before `MIN_TIME` it would be too early, but it works with whatever value in the window. Kind of like this image from stm32 datasheet. Is it different for samd21? Am I missing something obvious?

![image](https://user-images.githubusercontent.com/23060007/59385209-2942cb00-8d64-11e9-9337-af1da60f7cf0.png)

fjmolinas replied:
Maybe the "drawing" is confusing?

benpicco replied:
I agree that it should be `xtimer_usleep(MIN_TIME)` - but it's still `xtimer_usleep(VALID_TIME)` in the documentation.

fjmolinas replied:
Ok, I see, the I think it comes to the drawing. I was using `VALID_TIME` as  `MIN_TIME < VALID_TIME < MAX_TIME`, but the drawing makes it look as if it is `0 < VALID_TIME < MIN_TIME `. Maybe something like this would be more clear?

```
 *
 *
 * 0(ms)                 MIN_TIME(ms)                MAX_TIME(ms)
 *  |-------------------------|-----------WINDOW----------|
 *
 * 0(ms)                        VALID_TIME(ms)
 *  |---------------------------------|
 *                                    ^
 *                                    |
 *                               wdt_kick()
```

benpicco replied:
But `VALID_TIME` should not start before `MIN_TIME` is over.

I think the confusion stems from me thinking of `VALID_TIME` as a period that starts at `MIN_TIME` and ends at `MAX_TIME`. It's not a single point in time but a range.

fjmolinas replied:
Ok How about if we use your drawing and have: `xtimer_usleep(MIN_TIME+VALID_TIME)` ?

benpicco replied:
I'm fine with that. But having `VALID_TIME` there to me looks like it's a constant just like `MIN_TIME`, when it's not. It's `xtimer_usleep(MIN_TIME + x)`, where `x < MAX_TIME - MIN_TIME`.

fjmolinas replied:
Ok I think that clearer, thanks for the feedback :)

fjmolinas replied:
@benpicco I pushed according to your suggestions, does it look good now?

benpicco replied:
Just add another line of comment so we have `x` 'defined' in both cases and we're good. :)

benpicco commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,177 @@
+ *                           for NORMAL watchdog.
+ * @param[in] max_time       upper bound for WINDOWED watchdog in ms, time before
+ *                           reset for NORMAL watchdog.
+ * @param[in] cb _time       time before wdt_cb is executed, must.
```
I think this sentence should go on as

> must be < `max_time`, may be ignored if `min_time` > 0

benpicco commented:
I agree that `cb_time` is the more flexible API, but we should state in the documentation that it may be ignored by the implementation.

A `WDT_WARNING_PERIOD` would just shift the burden to a different place.

benpicco commented:
@kaspar030 @fjmolinas ping?

fjmolinas commented:
> I would also be fine with not having a configurable period and have the driver decide a default value.

I think this would easily lead to undocumented behavior.  Lets just go with `WDT_WARNING_PERIOD` which will be the time before the reset where the callback is enforced. For STM32, when implemented it can be asserted since the user can't define this.

@benpicco @kaspar030 If that is ok with you guys I'll change the PR and i documentation and we can move forward :)

kaspar030 commented:
> If that is ok with you guys I'll change the PR and update the documentation and we can move forward :)

ok!

fjmolinas commented:
Pushed the changes!

benpicco commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,199 @@
+ *           If WINDOWED set min_time and max_time
+ *
+ * @param[in] min_time       lower bound for WINDOWED watchdog in ms, has to be 0
+ *                           for NORMAL watchdog operation
```
time is in ms

benpicco commented:
Looks good to me, only one nit in the comment/example for windowed mode.

benpicco commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,202 @@
+ *
+ *    wdt_setup_reboot(0, MAX_TIME);
+ *    wdt_start();
+ * 
```
```
 *    // x <= MAX_TIME
```

aabadie commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,202 @@
+ *    wdt_setup_reboot(MIN_TIME, MAX_TIME);
+ *    wdt_start();
+ *
+ *    // x < (MAX_TIME - MIN_TIME)
```
Please use C style comments also in documentation.

I think code snippet must be in specific code blocks:
```
'''c
/* C code (replace ticks by back ticks) */
'''
```

fjmolinas replied:
If I put C style comments in a Doyxgen comment block it makes the whole code looks weird, I rather wither since it thinks the Doxygen comment block is ending there, so everything renders bad on IDE's

aabadie replied:
That's why you need to put the snippet in a code block. There's a Markdown syntax for this.

fjmolinas replied:
That doesn't work, RIOT code examples allways use "//" for comments within comments, you actually get compiler warning when compiling a file with "/* */" within another comment block

fjmolinas replied:
Also I saw a lot of code uses this for enclosing code blocks:

```
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
...
 * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Is markdown better?

aabadie replied:
I see and that's ok like this. But this is crap. Are the code comments really needed ?

fjmolinas replied:
I guess I can move the comment out of the code snippet to have this:

```
 * In the code snippet and diagram ```time``` is an arbitrary value such that
 * ```time < MAX_TIME```.
 *
 * @verbatim
 * 0(ms)                                            MAX_TIME(ms)
 * |----------------------------------------------------|
 *
 *                          time(ms)
 * |---------------------------| - - - - - - -  - - - - |
 *
 *                             ^
 *                             |
 *                        wdt_kick()
 * @endverbatim
 * 
 * ~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
 * #include <stdio.h>
 *
 * #include "periph/wdt.h"
 * #include "xtimer.h"
 *
 * int main(void)
 * {
 *     wdt_setup_reboot(0, MAX_TIME);
 *     wdt_start();
 *
 *     while (1) {
 *         xtimer_usleep(time);
 *         wdt_kick();
 *     }
 *     return 0;
 * }
 * ~~~~~~~~~~~~~~~~~~~~~~~~
```
 Would that also be ok with you @benpicco ? (I changed x to time)

benpicco replied:
Looks good to me.

fjmolinas commented:
Had to change the documentation a little since some code wasn't rendering properly in doxygen.

fjmolinas commented:
@kaspar030 @benpicco @aabadie pushed final documentation changes, I think everything should be OK, can you give it a (hopefully) final look?

aabadie commented on drivers/include/periph/wdt.h:
````diff
@@ -0,0 +1,290 @@
+ *
+ * A watchdog timer (WDT) is an electronic or software timer that is
+ * used to detect and recover from unusual or suspect behaviour as well as
+ * software malfunctions. During ```NORMAL``` operation the application will reset
````
You can use simple back ticks instead of triple backticks. There are quite a few to change, sorry ;)

aabadie commented:
@kaspar030, are you ok with the changes of this PR ?

aabadie commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,290 @@
+/**
+ * @def     WWDT_TIME_LOWER_LIMIT
+ *
+ * @brief   Lower limit in ms for wdt operating in NORMAL mode
```
Shouldn't this be the lower limit for **window** mode ?

fjmolinas commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,290 @@
+/**
+ * @def     NWDT_TIME_UPPER_LIMIT
+ *
+ * @brief   Upper limit in ms for wdt operating in WINDOW mode
```
yep addressed.

aabadie commented on drivers/include/periph/wdt.h:
```diff
@@ -0,0 +1,290 @@
+ * used to detect and recover from unusual or suspect behaviour as well as
+ * software malfunctions. During `NORMAL` operation the application will reset
+ * wdt_kick() the timer counter preventing it from hitting 0. If due to
+ * software or hardware error the WDT counter isn't reset this will trigger
```
Here I would replace `this` by `it` and maybe add a comma before `it` (`it` means the watchdog in this case.).

benpicco commented:
I think it makes sense to also have a `void wdt_init(void)` that can be called by auto_init, like we have for the other peripheral drivers.

E.g. if we want to re-configure the watchdog at run-time, we don't want to do the clock setup and such again.

kaspar030 commented:
> E.g. if we want to re-configure the watchdog at run-time, we don't want to do the clock setup and such again.

Would that hurt? Are we expecting to re-configure the wdt in a loop?

benpicco commented:
It would only hurt aesthetics - if you don't want to have a dedicated init function, I can live with that. Just think it would make the setup function(s) cleaner.

fjmolinas commented:
> It would only hurt aesthetics - if you don't want to have a dedicated init function, I can live with that. Just think it would make the setup function(s) cleaner.

I do see a use cases for changing the `wdt_setup_reboot` multiple times (that is why I think its good we can change the time values in `wdt_setup_reboot`). e.g: platforms where once enabled it can't be disabled, you might re-configure it to the highest value before going into sleep modes, in order to reduce the amount of wake-ups (just to reset the timer.) 

But it seems this can be really platform specific (looking at https://github.com/RIOT-OS/RIOT/pull/11556 and https://github.com/RIOT-OS/RIOT/pull/7374 ,`nrf5x`, `c2538` don't seem to need it and `stm32` don't either). Could this also be done in for e.g. samd clock initialization code? If `MODULES_PERIPH_WDT` is defined?

benpicco commented:
[configure watchdog behavior during sleep/debug](https://github.com/RIOT-OS/RIOT/pull/11556/files#diff-e4c193978f2684bb1325916a5d9b409aR75) could also be called in `init()`, it's only needed once.

I can also add an `is_initialized` variable if we don't want to do the clock setup multiple times and don't want to have empty init functions on some platforms.

fjmolinas commented:
> configure watchdog behavior during sleep/debug could also be called in init(), it's only needed once.

True, I should not have been only looking for clock configuration. I think you are right, there are multiple things that can be setup only once (clocks, behavior in sleep modes or debug, as you pointed out, etc..). There would still be some platforms that don't use it (but in a way same goes for `wdt_stop()`), but probably will.

> I can also add an `is_initialized` variable if we don't want to do the clock setup multiple times and don't want to have empty init functions on some platforms.

I don't have a technical argument against empty functions (although this can be my own ignorance :) ).

kaspar030 commented:
> I don't have a technical argument against empty functions (although this can be my own ignorance :) ).

We can also guard the call to ```wdt_init()``` with ```#ifdef HAS_WDT_INIT```.

benpicco commented:
> We can also guard the call to wdt_init() with #ifdef HAS_WDT_INIT.

So save maybe two instructions in ROM?
I'm all for being thrifty in regards to space, but I feel like this is pushing it a bit.

kaspar030 commented:
> So save maybe two instructions in ROM?
> I'm all for being thrifty in regards to space, but I feel like this is pushing it a bit.

We get this argument *all the time*, like on hundreds of occasions, "why save those two bytes"?

Each individual two bytes (in this case it would be at least 4 on ARM) really do not matter. But all the hundreds of times we decide to not care do add up. In the end, there's kilobytes of "n<$much bytes do not matter" in the final binary.

Of course this kind of optimization should not go too far. But IMO a centralized ifdef guard is acceptable.

benpicco commented:
OK, I'm fine with this.

aabadie commented:
So what's the conclusion with the init function ? Personally I don't see it very useful. Not every peripheral are automatically initialized in `periph_init`. Maybe this could be done by the auto_init module but that doesn't require a new init function in the API: the `auto_init_wdt` would just can `wdt_setup` with predefined parameters (passed via macros for example).

kaspar030 commented:
> So what's the conclusion with the init function ?

I'd suggest we keep it out of the API (it would be called automatically, anyways). The first implementation that wants to make use of it can then introduce it.

fjmolinas commented:
> I'd suggest we keep it out of the API (it would be called automatically, anyways). The first implementation that wants to make use of it can then introduce it.

In that case are we good with the PR as it is?

kaspar030 commented:
> In that case are we good with the PR as it is?

I think it is. Would it make sense to get implementation experience first? IIRC there are two wdt implementations "in the queue". Are they close to implementing this PR?

aabadie commented:
> there are two wdt implementations "in the queue"

in fact, there are 3: nrf5x, stm32 and sam0. I'll adapt the one for nrf.

fjmolinas commented:
> Each individual two bytes (in this case it would be at least 4 on ARM) really do not matter. But all the hundreds of times we decide to not care do add up. In the end, there's kilobytes of "n<$much bytes do not matter" in the final binary.

In that case I think `wdt_stop` should also be guarded. 

> I think it is. Would it make sense to get implementation experience first? IIRC there are two wdt implementations "in the queue". Are they close to implementing this PR?

I adapted my PR.

kaspar030 commented:
> In that case I think `wdt_stop` should also be guarded.

Hmm, not sure. In the "wdt_init()" case, we'd (conditionally) call the function in "periph_init()", if the module periph_wdt is compiled in. Only that call needs to be guarded.

For "wdt_stop()", if the application doesn't call it, it won't use any space in the binary.

fjmolinas commented:
> Hmm, not sure. In the "wdt_init()" case, we'd (conditionally) call the function in "periph_init()", if the module periph_wdt is compiled in. Only that call needs to be guarded.
> 
> For "wdt_stop()", if the application doesn't call it, it won't use any space in the binary.

Maybe but that would make coding generic code harder, If it is guarded against a `FEATURE` then in my application I can use only if the `FEATURE` is supported instead of a compilation warning for an un-implemented function, or an empty function that hides it is actually doing nothing (even if there is a debug message).

kaspar030 commented:
> Maybe but that would make coding generic code harder, If it is guarded against a `FEATURE` then in my application I can use only if the `FEATURE` is supported instead of a compilation warning for an un-implemented function, or an empty function that hides it is actually doing nothing (even if there is a debug message).

Hm, I don't understand your concern.

```wdt_stop()``` will only be compiled if the application add "periph/wdt" to "USEMODULE". Adding "periph/wdt" to "FEATURES_OPTIONAL" or "FEATURES_REQUIRED" does that implicitly. (If the target board doesn't support the feature, USEMODULE or FEATURES_REQUIRED" will fail at build time, FEATURES_OPTIONAL will just not add periph_wdt to USEMODULE.).
At that point, an application can use the wdt_* functions. The linker will ensure that unused functions won't enlarge the binary.

If an application expects to be run on both boards supporting and boards not supporting periph_wdt, it can add "periph_wdt" to "FEATRURES_OPTIONAL", but it has to take care of guarding the wdt_* calls with ```#ifdef MODULE_PERIPH_WDT```.
We don't do empty functions if possible, as they enlarge binaries.

For ```wdt_init()```, the way we do it with other periph init functions is to always call them in periph_init() (which is shared among *all* platforms). In order to not require an empty function, we introduce those extra guards like "PERIPH_WDT_HAS_INIT".

fjmolinas commented:
> wdt_stop() will only be compiled if the application add "periph/wdt" to "USEMODULE". Adding "periph/wdt" to "FEATURES_OPTIONAL" or "FEATURES_REQUIRED" does that implicitly. (If the target board doesn't support the feature, USEMODULE or FEATURES_REQUIRED" will fail at build time, FEATURES_OPTIONAL will just not add periph_wdt to USEMODULE.).
> At that point, an application can use the wdt_* functions. The linker will ensure that unused functions won't enlarge the binary.

I'll try to explain what I meant again. For many platforms`wdt` can't be stopped, once enabled it runs until a system reset, so `wdt_stop()` does nothing. A use case examples is an application where I would want to turn-off the `wdt` before entering sleep mode. If possible that means I can just go to sleep and start it again on wake-up. But if I can't really stop the `wdt` then I must implement something(`rtc alarm`) that will wake-up just before the `wdt` ends, reset it, and the immediately go to sleep. In this case both platforms would be supporting `wdt` but one of them can `stop` it and the other not. 

In that case I could do something like:

```
 // somewhere else in the application for platforms with `wdt`
 wdt_setup_reboot(0, MAX_TIME);
 wdt_start();

#ifdef MODULE_WDT_STOP
    wdt_stop()
#else 
   // a callback that will reset the counter, etc...
   rtc_set_alarm()
#endif
```

kaspar030 commented:
> In that case I could do something like:

I see. Those platforms would just not implement wdt_stop(). You're right, something is necessary to let the application handle this. I suggest "#define HAS_WDT_STOP".

fjmolinas commented:
@kaspar030 I did as you suggested.

benpicco commented:
I don't want to delay this any further, but should we also add `WDT_HAS_INIT`/`wdt_init()` while we are at it?
I'm fine with introducing it with the actual implementation too.

fjmolinas commented:
> I don't want to delay this any further, but should we also add WDT_HAS_INIT/wdt_init() while we are at it?
> I'm fine with introducing it with the actual implementation too.

I think it does if there are queued PR's that make use of it right away, and it seems nrf52, stm32 and sam0 will or could make use of it. If we rebase our PR's to make use of it then I think we should include them in the headers.

fjmolinas commented:
> (feel free to add WDT_HAS_INIT here or where needed...)

I added `WDT_HAS_INIT`. @kaspar030 Can I squash?

kaspar030 commented:
>  @kaspar030 Can I squash?

yes please!

fjmolinas commented:
Squashed, I fixed an issue where `WDT_WARNING_PERIOD` group was wrongly defined and added a `@note` for `wdt_init`

kaspar030 commented:
&go!
